### PR TITLE
Fix : asr verify on external symbol

### LINKED ISSUE #3968: allocating using module variable and using module routine outside module throws error
> mms_copy.f90

```fortran
MODULE XX
  IMPLICIT NONE

  INTEGER, PUBLIC :: nx = 1000

END MODULE

MODULE mms_module

  USE XX, ONLY: nx

  IMPLICIT NONE
  REAL, ALLOCATABLE, DIMENSION(:), PUBLIC :: ref_flux

  REAL, ALLOCATABLE, DIMENSION(:), PUBLIC :: ref_fluxm

  CONTAINS

  SUBROUTINE mms_deallocate
  END SUBROUTINE mms_deallocate

  SUBROUTINE mms_allocate ( )
    ALLOCATE( ref_flux(nx), ref_fluxm(nx) )
  END SUBROUTINE mms_allocate
END MODULE mms_module
```

> dealloc.f90

```fortran
MODULE dealloc_module
  USE mms_module, ONLY: mms_deallocate
END MODULE dealloc_module
```

```console
% lfortran -c mms_copy.f90 && lfortran -c dealloc.f90
Internal Compiler Error: Unhandled exception
Traceback (most recent call last):
  File "/Users/pranavchiku/repos/lfortran/src/bin/lfortran.cpp", line 2459
    LCompilers::print_stack_on_segfault();
  File "/Users/pranavchiku/repos/lfortran/src/bin/lfortran.cpp", line 2391
    return compile_to_object_file(arg_file, outfile, false,
  File "/Users/pranavchiku/repos/lfortran/src/bin/lfortran.cpp", line 906
    result = fe.get_asr2(input, lm, diagnostics);
  File "/Users/pranavchiku/repos/lfortran/src/lfortran/fortran_evaluator.cpp", line 252
    Result<ASR::TranslationUnit_t*> res2 = get_asr3(*ast, diagnostics);
  File "/Users/pranavchiku/repos/lfortran/src/lfortran/fortran_evaluator.cpp", line 274
    compiler_options.symtab_only, compiler_options);
  File "/Users/pranavchiku/repos/lfortran/src/lfortran/semantics/ast_to_asr.cpp", line 77
    if (res.ok) {
  File "/Users/pranavchiku/repos/lfortran/src/lfortran/semantics/ast_symboltable_visitor.cpp", line 3457
    v.visit_TranslationUnit(ast);
  File "/Users/pranavchiku/repos/lfortran/src/lfortran/semantics/ast_symboltable_visitor.cpp", line 120
    visit_ast(*x.m_items[i]);
  File "/Users/pranavchiku/repos/lfortran/src/lfortran/ast.h", line 4878
    void visit_ast(const ast_t &b) { visit_ast_t(b, self()); }
  File "/Users/pranavchiku/repos/lfortran/src/lfortran/ast.h", line 4834
    case astType::unit: { v.visit_unit((const unit_t &)x); return; }
  File "/Users/pranavchiku/repos/lfortran/src/lfortran/ast.h", line 4881
    void visit_mod(const mod_t &b) { visit_mod_t(b, self()); }
  File "/Users/pranavchiku/repos/lfortran/src/lfortran/ast.h", line 4465
    case modType::Module: { v.visit_Module((const Module_t &)x); return; }
  File "/Users/pranavchiku/repos/lfortran/src/lfortran/semantics/ast_symboltable_visitor.cpp", line 471
    visit_ModuleSubmoduleCommon<AST::Module_t, ASR::Module_t>(x);
  File "/Users/pranavchiku/repos/lfortran/src/lfortran/semantics/ast_symboltable_visitor.cpp", line 424
    visit_unit_decl1(*x.m_use[i]);
  File "/Users/pranavchiku/repos/lfortran/src/lfortran/ast.h", line 4890
    void visit_unit_decl1(const unit_decl1_t &b) { visit_unit_decl1_t(b, self()); }
  File "/Users/pranavchiku/repos/lfortran/src/lfortran/ast.h", line 4486
    case unit_decl1Type::Use: { v.visit_Use((const Use_t &)x); return; }
  File "/Users/pranavchiku/repos/lfortran/src/lfortran/semantics/ast_symboltable_visitor.cpp", line 2565
    t = (ASR::symbol_t*)(ASRUtils::load_module(al, tu_symtab,
  File "/Users/pranavchiku/repos/lfortran/src/libasr/asr_utils.cpp", line 308
    *symtab, intrinsic, pass_options);
  File "/Users/pranavchiku/repos/lfortran/src/libasr/asr_utils.cpp", line 456
    ASR::TranslationUnit_t *asr = load_modfile(al, modfile, false, symtab);
  File "/Users/pranavchiku/repos/lfortran/src/libasr/modfile.cpp", line 93
    ASR::asr_t *asr = deserialize_asr(al, asr_binary, load_symtab_id, symtab);
  File "/Users/pranavchiku/repos/lfortran/src/lfortran/ast_serialization.cpp", line 414
    Line not found
  File "/Users/pranavchiku/repos/lfortran/src/lfortran/ast_serialization.cpp", line 430
    Line not found
  File "/Users/pranavchiku/repos/lfortran/src/libasr/asr_verify.cpp", line 1242
    v.visit_TranslationUnit(unit);
  File "/Users/pranavchiku/repos/lfortran/src/libasr/asr_verify.cpp", line 111
    for (auto &a : x.m_symtab->get_scope()) {
  File "../libasr/asr.h", line 5143
  File "../libasr/asr.h", line 4855
  File "/Users/pranavchiku/repos/lfortran/src/libasr/asr_verify.cpp", line 333
    for (auto &a : x.m_symtab->get_scope()) {
  File "../libasr/asr.h", line 5143
  File "../libasr/asr.h", line 4856
  File "/Users/pranavchiku/repos/lfortran/src/libasr/asr_verify.cpp", line 454
    visit_stmt(*x.m_body[i]);
  File "../libasr/asr.h", line 5160
  File "../libasr/asr.h", line 4878
  File "/Users/pranavchiku/repos/lfortran/src/libasr/asr_verify.cpp", line 1218
    BaseWalkVisitor<VerifyVisitor>::visit_Allocate(x);
  File "../libasr/asr.h", line 5477
  File "../libasr/asr.h", line 6619
  File "/Users/pranavchiku/repos/lfortran/src/libasr/asr_verify.cpp", line 1171
    require_with_loc(ASRUtils::is_integer(
  File "../libasr/asr_utils.h", line 294
  File "../libasr/asr.h", line 50299
AssertFailed: e->m_external
```

### LINKED ISSUE #3977: AssertFailed: e->m_external
```console
$ make
lfortran  -c global.f90 -o global.o
lfortran  -c sn.f90 -o sn.o
lfortran  -c data.f90 -o data.o
lfortran  -c geom.f90 -o geom.o
lfortran  -c control.f90 -o control.o
cpp -P  time.0.F90 >time.f90
lfortran  -c time.f90
warning: Assuming implicit save attribute for variable declaration
  --> time.f90:60:25
   |
60 |     CHARACTER(LEN=1) :: star='*'
   |                         ^^^^^^^^ help: add explicit save attribute or initialize in a separate statement


Note: Please report unclear, confusing or incorrect messages as bugs at
https://github.com/lfortran/lfortran/issues.
cpp -P  plib.0.F90 >plib.f90
lfortran  -c plib.f90
lfortran  -c mms.f90 -o mms.o
warning: Assuming implicit save attribute for variable declaration
   --> mms.f90:522:25
    |
522 |     CHARACTER(LEN=1) :: star='*'
    |                         ^^^^^^^^ help: add explicit save attribute or initialize in a separate statement


Note: Please report unclear, confusing or incorrect messages as bugs at
https://github.com/lfortran/lfortran/issues.
lfortran  -c solvar.f90 -o solvar.o
lfortran  -c dealloc.f90 -o dealloc.o
Internal Compiler Error: Unhandled exception
BFD: DWARF error: invalid or unhandled FORM value: 0x23
BFD: DWARF error: invalid or unhandled FORM value: 0x23
BFD: DWARF error: invalid or unhandled FORM value: 0x23
BFD: DWARF error: invalid or unhandled FORM value: 0x23
BFD: DWARF error: invalid or unhandled FORM value: 0x23
BFD: DWARF error: invalid or unhandled FORM value: 0x23
BFD: DWARF error: invalid or unhandled FORM value: 0x23
BFD: DWARF error: invalid or unhandled FORM value: 0x23
BFD: DWARF error: invalid or unhandled FORM value: 0x23
BFD: DWARF error: invalid or unhandled FORM value: 0x23
BFD: DWARF error: invalid or unhandled FORM value: 0x23
BFD: DWARF error: invalid or unhandled FORM value: 0x23
BFD: DWARF error: invalid or unhandled FORM value: 0x23
BFD: DWARF error: invalid or unhandled FORM value: 0x23
BFD: DWARF error: invalid or unhandled FORM value: 0x23
BFD: DWARF error: invalid or unhandled FORM value: 0x23
BFD: DWARF error: invalid or unhandled FORM value: 0x23
BFD: DWARF error: invalid or unhandled FORM value: 0x23
BFD: DWARF error: invalid or unhandled FORM value: 0x23
BFD: DWARF error: invalid or unhandled FORM value: 0x23
BFD: DWARF error: invalid or unhandled FORM value: 0x23
BFD: DWARF error: invalid or unhandled FORM value: 0x23
BFD: DWARF error: invalid or unhandled FORM value: 0x23
BFD: DWARF error: invalid or unhandled FORM value: 0x23
BFD: DWARF error: invalid or unhandled FORM value: 0x23
BFD: DWARF error: invalid or unhandled FORM value: 0x23
BFD: DWARF error: invalid or unhandled FORM value: 0x23
BFD: DWARF error: invalid or unhandled FORM value: 0x23
BFD: DWARF error: invalid or unhandled FORM value: 0x23
BFD: DWARF error: invalid or unhandled FORM value: 0x23
BFD: DWARF error: invalid or unhandled FORM value: 0x23
BFD: DWARF error: invalid or unhandled FORM value: 0x23
BFD: DWARF error: invalid or unhandled FORM value: 0x23
BFD: DWARF error: invalid or unhandled FORM value: 0x23
BFD: DWARF error: invalid or unhandled FORM value: 0x23
BFD: DWARF error: invalid or unhandled FORM value: 0x23
BFD: DWARF error: invalid or unhandled FORM value: 0x23
BFD: DWARF error: invalid or unhandled FORM value: 0x23
BFD: DWARF error: invalid or unhandled FORM value: 0x23
BFD: DWARF error: invalid or unhandled FORM value: 0x23
BFD: DWARF error: invalid or unhandled FORM value: 0x23
BFD: DWARF error: invalid or unhandled FORM value: 0x23
Traceback (most recent call last):
  Binary file "/home/ondrej/repos/lfortran/lfortran/src/bin/lfortran", in _start()
  Binary file "/lib/x86_64-linux-gnu/libc.so.6", in __libc_start_main()
  Binary file "/lib/x86_64-linux-gnu/libc.so.6", in __libc_init_first()
  Binary file "/home/ondrej/repos/lfortran/lfortran/src/bin/lfortran", in main()
  Binary file "/home/ondrej/repos/lfortran/lfortran/src/bin/lfortran", in main_app(int, char**)
  File "lfortran.cpp", line 0, in (anonymous namespace)::compile_to_object_file(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool, LCompilers::CompilerOptions&, LCompilers::PassManager&)
  Binary file "/home/ondrej/repos/lfortran/lfortran/src/bin/lfortran", in LCompilers::FortranEvaluator::get_asr2(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, LCompilers::LocationManager&, LCompilers::diag::Diagnostics&)
  Binary file "/home/ondrej/repos/lfortran/lfortran/src/bin/lfortran", in LCompilers::FortranEvaluator::get_asr3(LCompilers::LFortran::AST::TranslationUnit_t&, LCompilers::diag::Diagnostics&)
  Binary file "/home/ondrej/repos/lfortran/lfortran/src/bin/lfortran", in LCompilers::LFortran::ast_to_asr(Allocator&, LCompilers::LFortran::AST::TranslationUnit_t&, LCompilers::diag::Diagnostics&, LCompilers::SymbolTable*, bool, LCompilers::CompilerOptions&)
  Binary file "/home/ondrej/repos/lfortran/lfortran/src/bin/lfortran", in LCompilers::LFortran::symbol_table_visitor(Allocator&, LCompilers::LFortran::AST::TranslationUnit_t&, LCompilers::diag::Diagnostics&, LCompilers::SymbolTable*, LCompilers::CompilerOptions&, std::map<unsigned long, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, LCompilers::ASR::ttype_t*, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, LCompilers::ASR::ttype_t*> > >, std::less<unsigned long>, std::allocator<std::pair<unsigned long const, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, LCompilers::ASR::ttype_t*, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, LCompilers::ASR::ttype_t*> > > > > >&, std::map<unsigned long, LCompilers::ASR::symbol_t*, std::less<unsigned long>, std::allocator<std::pair<unsigned long const, LCompilers::ASR::symbol_t*> > >&, std::map<unsigned long, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::less<unsigned long>, std::allocator<std::pair<unsigned long const, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > > >&, std::map<unsigned int, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, LCompilers::ASR::ttype_t*, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, LCompilers::ASR::ttype_t*> > >, std::less<unsigned int>, std::allocator<std::pair<unsigned int const, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, LCompilers::ASR::ttype_t*, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, LCompilers::ASR::ttype_t*> > > > > >&, std::map<unsigned int, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, LCompilers::ASR::symbol_t*, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, LCompilers::ASR::symbol_t*> > >, std::less<unsigned int>, std::allocator<std::pair<unsigned int const, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, LCompilers::ASR::symbol_t*, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, LCompilers::ASR::symbol_t*> > > > > >&, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::vector<LCompilers::LFortran::AST::stmt_t*, std::allocator<LCompilers::LFortran::AST::stmt_t*> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::vector<LCompilers::LFortran::AST::stmt_t*, std::allocator<LCompilers::LFortran::AST::stmt_t*> > > > >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::vector<LCompilers::LFortran::AST::stmt_t*, std::allocator<LCompilers::LFortran::AST::stmt_t*> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::vector<LCompilers::LFortran::AST::stmt_t*, std::allocator<LCompilers::LFortran::AST::stmt_t*> > > > > > > >&, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::vector<int, std::allocator<int> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::vector<int, std::allocator<int> > > > >&, std::vector<LCompilers::ASR::stmt_t*, std::allocator<LCompilers::ASR::stmt_t*> >&)
  Binary file "/home/ondrej/repos/lfortran/lfortran/src/bin/lfortran", in LCompilers::LFortran::SymbolTableVisitor::visit_TranslationUnit(LCompilers::LFortran::AST::TranslationUnit_t const&)
  Binary file "/home/ondrej/repos/lfortran/lfortran/src/bin/lfortran", in LCompilers::LFortran::AST::BaseVisitor<LCompilers::LFortran::SymbolTableVisitor>::visit_ast(LCompilers::LFortran::AST::ast_t const&)
  File "ast_symboltable_visitor.cpp", line 0, in void LCompilers::LFortran::AST::visit_ast_t<LCompilers::LFortran::SymbolTableVisitor>(LCompilers::LFortran::AST::ast_t const&, LCompilers::LFortran::SymbolTableVisitor&)
  Binary file "/home/ondrej/repos/lfortran/lfortran/src/bin/lfortran", in LCompilers::LFortran::AST::BaseVisitor<LCompilers::LFortran::SymbolTableVisitor>::visit_mod(LCompilers::LFortran::AST::mod_t const&)
  File "ast_symboltable_visitor.cpp", line 0, in void LCompilers::LFortran::AST::visit_mod_t<LCompilers::LFortran::SymbolTableVisitor>(LCompilers::LFortran::AST::mod_t const&, LCompilers::LFortran::SymbolTableVisitor&)
  Binary file "/home/ondrej/repos/lfortran/lfortran/src/bin/lfortran", in LCompilers::LFortran::SymbolTableVisitor::visit_Module(LCompilers::LFortran::AST::Module_t const&)
  Binary file "/home/ondrej/repos/lfortran/lfortran/src/bin/lfortran", in void LCompilers::LFortran::SymbolTableVisitor::visit_ModuleSubmoduleCommon<LCompilers::LFortran::AST::Module_t, LCompilers::ASR::Module_t>(LCompilers::LFortran::AST::Module_t const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)
  Binary file "/home/ondrej/repos/lfortran/lfortran/src/bin/lfortran", in LCompilers::LFortran::AST::BaseVisitor<LCompilers::LFortran::SymbolTableVisitor>::visit_unit_decl1(LCompilers::LFortran::AST::unit_decl1_t const&)
  File "ast_symboltable_visitor.cpp", line 0, in void LCompilers::LFortran::AST::visit_unit_decl1_t<LCompilers::LFortran::SymbolTableVisitor>(LCompilers::LFortran::AST::unit_decl1_t const&, LCompilers::LFortran::SymbolTableVisitor&)
  Binary file "/home/ondrej/repos/lfortran/lfortran/src/bin/lfortran", in LCompilers::LFortran::SymbolTableVisitor::visit_Use(LCompilers::LFortran::AST::Use_t const&)
  Binary file "/home/ondrej/repos/lfortran/lfortran/src/bin/lfortran", in LCompilers::ASRUtils::load_module(Allocator&, LCompilers::SymbolTable*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, LCompilers::Location const&, bool, LCompilers::PassOptions&, bool, std::function<void (std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, LCompilers::Location const&)>)
  Binary file "/home/ondrej/repos/lfortran/lfortran/src/bin/lfortran", in LCompilers::ASRUtils::find_and_load_module(Allocator&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, LCompilers::SymbolTable&, bool, LCompilers::PassOptions&)
  Binary file "/home/ondrej/repos/lfortran/lfortran/src/bin/lfortran", in LCompilers::load_modfile(Allocator&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool, LCompilers::SymbolTable&)
  Binary file "/home/ondrej/repos/lfortran/lfortran/src/bin/lfortran", in LCompilers::deserialize_asr(Allocator&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool, LCompilers::SymbolTable&)
  Binary file "/home/ondrej/repos/lfortran/lfortran/src/bin/lfortran", in LCompilers::deserialize_asr(Allocator&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool)
  Binary file "/home/ondrej/repos/lfortran/lfortran/src/bin/lfortran", in LCompilers::asr_verify(LCompilers::ASR::TranslationUnit_t const&, bool, LCompilers::diag::Diagnostics&)
  Binary file "/home/ondrej/repos/lfortran/lfortran/src/bin/lfortran", in LCompilers::ASR::VerifyVisitor::visit_TranslationUnit(LCompilers::ASR::TranslationUnit_t const&)
  Binary file "/home/ondrej/repos/lfortran/lfortran/src/bin/lfortran", in LCompilers::ASR::BaseVisitor<LCompilers::ASR::VerifyVisitor>::visit_symbol(LCompilers::ASR::symbol_t const&)
  File "asr_verify.cpp", line 0, in void LCompilers::ASR::visit_symbol_t<LCompilers::ASR::VerifyVisitor>(LCompilers::ASR::symbol_t const&, LCompilers::ASR::VerifyVisitor&)
  Binary file "/home/ondrej/repos/lfortran/lfortran/src/bin/lfortran", in LCompilers::ASR::VerifyVisitor::visit_Module(LCompilers::ASR::Module_t const&)
  Binary file "/home/ondrej/repos/lfortran/lfortran/src/bin/lfortran", in LCompilers::ASR::BaseVisitor<LCompilers::ASR::VerifyVisitor>::visit_symbol(LCompilers::ASR::symbol_t const&)
  File "asr_verify.cpp", line 0, in void LCompilers::ASR::visit_symbol_t<LCompilers::ASR::VerifyVisitor>(LCompilers::ASR::symbol_t const&, LCompilers::ASR::VerifyVisitor&)
  Binary file "/home/ondrej/repos/lfortran/lfortran/src/bin/lfortran", in LCompilers::ASR::VerifyVisitor::visit_Function(LCompilers::ASR::Function_t const&)
  Binary file "/home/ondrej/repos/lfortran/lfortran/src/bin/lfortran", in LCompilers::ASR::BaseVisitor<LCompilers::ASR::VerifyVisitor>::visit_stmt(LCompilers::ASR::stmt_t const&)
  File "asr_verify.cpp", line 0, in void LCompilers::ASR::visit_stmt_t<LCompilers::ASR::VerifyVisitor>(LCompilers::ASR::stmt_t const&, LCompilers::ASR::VerifyVisitor&)
  Binary file "/home/ondrej/repos/lfortran/lfortran/src/bin/lfortran", in LCompilers::ASR::VerifyVisitor::visit_Allocate(LCompilers::ASR::Allocate_t const&)
  Binary file "/home/ondrej/repos/lfortran/lfortran/src/bin/lfortran", in LCompilers::ASR::BaseWalkVisitor<LCompilers::ASR::VerifyVisitor>::visit_Allocate(LCompilers::ASR::Allocate_t const&)
  Binary file "/home/ondrej/repos/lfortran/lfortran/src/bin/lfortran", in LCompilers::ASR::BaseWalkVisitor<LCompilers::ASR::VerifyVisitor>::visit_alloc_arg(LCompilers::ASR::alloc_arg_t const&)
  Binary file "/home/ondrej/repos/lfortran/lfortran/src/bin/lfortran", in LCompilers::ASR::VerifyVisitor::visit_dimension(LCompilers::ASR::dimension_t const&)
  File "asr_verify.cpp", line 0, in LCompilers::ASRUtils::expr_type(LCompilers::ASR::expr_t const*)
AssertFailed: e->m_external
make: *** [Makefile:147: dealloc.o] Error 1
```

Possibly a duplicate of #3968.

### PR DESCRIPTION
**Pull request by assem2002**
closes #3968 ,closes #3977 
The code base is working fine.
 
The problem rises from running `asr_verify()` on `ASR::TranslationUnit` node that has its `ExternalSymbol` nodes not completely built as the `FixExternalSymbol` class start its work after running `asr_verify()`.

**THE FIX** : We would just memoize all `ASR::TranslationUnit` and do `FixExternalSymbol` then finally run `asr_verify` on all of them.
Though we run `asr_verify` on the whole `ASR::TranslationUnit` after inserting all external modules in the main symtab, but it's a good practice to verify every single module we deserialize one by one.

**Visualization :)** 

Screenshot from 2024-06-05 23-32-53

Screenshot from 2024-06-05 23-40-28

### CODE REVIEW

File: src/libasr/serialization.cpp [hunk 1]
```diff
@@ -424,14 +424,8 @@ ASR::asr_t* deserialize_asr(Allocator &al, const std::string &s,
...
     ASR::FixParentSymtabVisitor p;
     p.visit_TranslationUnit(*tu);
-
-#if defined(WITH_LFORTRAN_ASSERT)
-    diag::Diagnostics diagnostics;
-    if (!asr_verify(*tu, false, diagnostics)) {
-        std::cerr << diagnostics.render2();
-        throw LCompilersException("Verify failed");
-    };
- #endif
```
Comment (certik) [inline src/libasr/serialization.cpp:434]: We need to keep this, to ensure the loaded code passes verify.

File: src/libasr/asr_utils.cpp [hunk 2]
```diff
@@ -387,6 +390,16 @@ ASR::Module_t* load_module(Allocator &al, SymbolTable *symtab,
...
+        for(auto &mod_to_verify : all_imported_mod_files){
+            ASR::TranslationUnit_t* trans_node = mod_to_verify.first;
+            ASR::Module_t *inbodied_mod = extract_module(*trans_node);
+            inbodied_mod->m_symtab->parent = trans_node->m_symtab; // set parent of module to its original TranslationUnit symtab.
+            if (!asr_verify(*trans_node, false, diagnostics)) {
+            std::cerr << diagnostics.render2();
+            throw LCompilersException("Verify failed in mod file " + mod_to_verify.second);
+            };
+            inbodied_mod->m_symtab->parent = symtab; // set parent back to the symtab we're currently building.
+        }
```
Comment (certik) [inline src/libasr/asr_utils.cpp:402]: Why was this moved here from the deserialization?

**Comment by certik** [comment]
The way it works is as follows:
* file gets deserialized and *immediately* checked with verify, and an option is passed in to skip following External nodes (the pointer will be nullptr). This checks everything and catches all deserialization bugs, except possibly external symbols.
* We construct the full ASR
* We run verify again, this time including all external nodes, so everything checks out.

In your alternative design, you skip the first step, and only check it once ASR is build/resolved. That seems better than the last step, but worse than the first step. Why not doing this in addition to the previous design? Or not doing it at all, since I think the current design should catch all bugs?

File: src/libasr/serialization.cpp (hunk 1 above)
Comment (assem2002) [reply src/libasr/serialization.cpp:434]: We do it later on just after we fix ExternalSymbol.
We can say I didn't remove, I've just postponed its effect.

Does the `TranslationUnit` node returning from `derserialize_asr()` get changed when building the main ASR?

**Comment by assem2002** [comment]
Yeah I got that.It seems that I passed a critical check point (the first asr_verify before fixing ExternalSymbol).

I thought we just needed to check every `deserialized_asr()` individually so I postponed the checks to be done after fixing ExternalSymbol.  
I don't get the benefit from verifying the asr before fixing its ExternalSymbol, I would like to know the reason behind it so maybe we can get into a middle solution.

We really needed to fix all ExternalSymbol because we have verification visitors that dive deep into the root of the ExternalSymbol to verify somethings. As example `visit_dimension`, it checks the variable to see whether its an integer or not.

**Comment by assem2002** [comment]
The main blocker of the issues mentioned in the PR is `visit_dimension` while verifying. 
As you mentioned `check_external` parameter, we can use it to bypass this verification point. Does that seem a better solution to solve this blocker?

**Comment by certik** [comment]
The reason we check ASR right after deserializing is to ensure things were constructed correctly. The goal of the ASR design should be such that this is possible. The only thing we can't check are `ExternalSymbol` because they point to other modules that must be resolved. I think it's a good design to check things right when they happen. 

Now let's investigate the problem at hand. Here is the `visit_dimension` verification: https://github.com/lfortran/lfortran/issues/3968#issuecomment-2151093422.

I can now see the issue, we can't resolve types right after deserializing. 

I currently don't know the best solution. In order to move forward, your solution is not bad. I am going to leave several comments above to clarify and understand it better.

File: src/libasr/asr_utils.cpp [hunk 3]
```diff
@@ -304,6 +304,7 @@ ASR::Module_t* load_module(Allocator &al, SymbolTable *symtab,
         }
     }
     LCOMPILERS_ASSERT(symtab->parent == nullptr);
+    std::vector<std::pair<ASR::TranslationUnit_t*,std::string>> all_imported_mod_files; // It holds all the mod files and their names, so we can apply asr_verify() on them later on.
```
Comment (certik) [inline src/libasr/asr_utils.cpp:307]: This should be guarded by `#if defined(WITH_LFORTRAN_ASSERT)`, since it is not needed otherwise, to ensure there is no performance penalty in release mode.

File: src/libasr/asr_utils.cpp [hunk 4]
```diff
@@ -319,6 +320,7 @@ ASR::Module_t* load_module(Allocator &al, SymbolTable *symtab,
         err("Module '" + module_name + "' not declared in the current source and the modfile was not found",
             loc);
     }
+    all_imported_mod_files.push_back({mod1,module_name});
```
Comment (certik) [inline src/libasr/asr_utils.cpp:323]: same here

File: src/libasr/asr_utils.cpp [hunk 5]
```diff
@@ -362,6 +364,7 @@ ASR::Module_t* load_module(Allocator &al, SymbolTable *symtab,
                 if (mod1 == nullptr) {
                     err("Module '" + item + "' modfile was not found", loc);
                 }
+                all_imported_mod_files.push_back({mod1,item});
```
Comment (certik) [inline src/libasr/asr_utils.cpp:367]: same here

File: src/libasr/asr_utils.cpp (hunk 2 above)
Comment (assem2002) [reply src/libasr/asr_utils.cpp:402]: You're right, It's my bad. I passed `false` to test something and forget to set it back to true

Comment (assem2002) [reply src/libasr/asr_utils.cpp:402]: Yes they're redundant. If we want to merge them together we can push_back the final `TranslationUnit` (`tu` in this code) and process them all and add if condition to handle the mutation of parent symtab

Comment (assem2002) [reply src/libasr/asr_utils.cpp:402]: If you see that verifying each imported file isn't important anymore  at this stage, We could just remove the loop.

Comment (certik) [reply src/libasr/asr_utils.cpp:402]: Can you see any advantage of running verify on each file in isolation, and then on the full ASR right after that?

I can see the advantage of running verify right after deserializing it. But at the stage above it seems redundant to me.

**Comment by Pranavchiku** [comment]
This PR does not have changes to incorporate `FixExternalSymbol`, right?

File: src/libasr/asr_utils.cpp (hunk 2 above)
Comment (assem2002) [reply src/libasr/asr_utils.cpp:402]: If you see that's is the case here. We would just turn the fix in this PR into just removing the `asr_verify()` call after deserializing each modfile. It that's the case just let me know and I'll push the new edits.

**Comment by assem2002** [comment]
> This PR does not have changes to incorporate `FixExternalSymbol`, right?

No, It doesn't.
